src/mhz16/mhz16.cxx

```diff
@@ -51,17 +50,44 @@ MHZ16::MHZ16(int uart)
     {
       throw std::runtime_error(std::string(__FUNCTION__) +
                                ": mraa_uart_get_dev_path() failed");
-      return;
     }
 
   // now open the tty
   if ( (m_ttyFd = open(devPath, O_RDWR)) == -1)
     {
       throw std::runtime_error(std::string(__FUNCTION__) +
-                               ": open of " + 
+                               ": open of " +
+                               string(devPath) + " failed: " +
+                               string(strerror(errno)));
+    }
+}
+
+MHZ16::MHZ16(const std::string& uart_raw)
+{
+  m_ttyFd = -1;
+
+  if ( !(m_uart = mraa_uart_init_raw(uart_raw.c_str())) )
+    {
+      throw std::invalid_argument(std::string(__FUNCTION__) +
+                                  ": mraa_uart_init_raw() failed");
+    }
+
+  // This requires a recent MRAA (1/2015)
+  const char *devPath = mraa_uart_get_dev_path(m_uart);
+
+  if (!devPath)
+    {
+      throw std::runtime_error(std::string(__FUNCTION__) +
+                               ": mraa_uart_get_dev_path() failed");
+    }
+
+  // now open the tty
+  if ( (m_ttyFd = open(devPath, O_RDWR)) == -1)
+    {
```

Comment:
Much of this stuff could be replaced with the current mraa_uart* functionality... When this code was written, MRAA did not have much of a UART API at all, hence the direct open(), setupTty(), dataAvailable() implementations, etc.  Maybe best saved for a later rewrite though.

Reply:
Agreed.  I would like to see more of the MRAA C++ used throughout the code base.